File: python/src/verifierloop_analysis/schemas/expectation.py
"""Expectation flag (mirror of crates/metrics/src/expectation.rs).

The "did the loop expect this?" field — measures confirmation bias, kept separate
from raw core observation.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Optional


class Expectation(str, Enum):
    """JSON values match the Rust `Expectation` enum (snake_case)."""

    NO_PREDICTION = "no_prediction"
    EXPECTED = "expected"
    UNEXPECTED = "unexpected"


@dataclass
class ExpectationFlag:
    expected: Expectation = Expectation.NO_PREDICTION
    predicted: Optional[str] = None

File: crates/metrics/src/expectation.rs
//! The "did the loop expect this?" flag — a dedicated field to MEASURE
//! confirmation bias, kept separate from raw core observation.

use serde::{Deserialize, Serialize};

/// Whether the loop predicted the observed outcome before seeing it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Expectation {
    /// The loop made no prediction (default).
    #[default]
    NoPrediction,
    /// The observation matched the loop's prediction.
    Expected,
    /// The observation contradicted the loop's prediction.
    Unexpected,
}

/// The confirmation-bias probe, kept separate from core observation.
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct ExpectationFlag {
    /// Expected / Unexpected / NoPrediction.
    pub expected: Expectation,
    /// What the loop predicted (if anything), for later bias analysis.
    pub predicted: Option<String>,
}
